### ISSUE TITLE
[SURE-6630] High CPU usage on Rancher cluster nodes when downstream clusters are shutdown

### ISSUE THREAD

**Issue by moio** [rancher/rancher#42864]
Internal reference: [SURE-6630](https://jira.suse.com/browse/SURE-6630)

**Rancher Server Setup**
- Rancher version: 2.7.5
- Installation option (Docker install/Helm Chart): Helm Chart, AKS

**Information about the Cluster**
- Kubernetes version: 1.26
- Cluster Type (Local/Downstream): 50 Downstream clusters (90% AKS and the remaining RKE1)

**User Information**
- What is the role of the user logged in? Irrelevant

**Describe the bug**
Some clusters are periodically unavailable, from a few to up to 55% of them. High CPU usage from rancher pods is noticed with one node always being at 100% CPU, and it worsens whenever a bigger number of clusters is not available.

A profile captured in a moment when most clusters were not available highlights the following hotspots:

![sure-6630_analysis](https://github.com/rancher/rancher/assets/250541/cfb4688f-5dc6-43e3-a9d7-17f036d929bd)
`go tool pprof -http=: ~/Downloads/after-profile-2023-09-08_16_03_45/rancher-5d6f96fdfd-ml58q-profile-2023-09-08T14_44_50`

**Hotspot 1** is the `restrictedAdminGlobalBindingsFleet` controller handler, which continuously rewrites `fleetworkspace` RoleBindings (2 per Restricted Admin, per cluster) even in absence of any changes.

 - when downstream clusters are available it fires every 5 minutes ([kubelet's nodeStatusReportFrequency](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) default)
 - when downstream clusters are not available it fires more often, every ~50s in my setup (30s + other compute time which depends on hardware ant other factors), every 2 minutes after the 30th time

It is believed this handler is also responsible for a significant amount of watch traffic and possibly memory trashing (GC time).

**To Reproduce**

I have not been able to reproduce all symptoms of this bug faithfully, however, **Hotspot 1** can easily be detected via `kubectl get --watch`.

Specifically:

 - set up a Rancher installation with at least one downstream cluster
 - set up at least one Restricted Admin user
 - run `kubectl get -A --watch --watch-only rolebinding`

(see later section for a full reproducer)

**Result**
 - `kubectl` outputs two `fleetworkspace` RoleBinding lines every 5 minutes (per Restricted Admin, per downstream cluster)

**Expected Result**
 - no output

**Additional context**

Full reproducer below:

```
# from a machine running Docker or with access to a Docker daemon
# with kubectl, helm, terraform, git and node installed

git clone -b disconnected_clusters https://github.com/moio/scalability-tests.git
cd scalability-tests
./bin/setup.mjs && ./bin/run_tests.mjs
KUBECONFIG=./config/upstream.yaml kubectl get -A --watch --watch-only rolebinding
```

Note: the reproducer uses the [disconnected_clusters branch](https://github.com/moio/scalability-tests/compare/master...disconnected_clusters) of the [scalability-tests project](https://github.com/moio/scalability-tests) which will deploy k3d clusters, install and set up Rancher, and create Restricted Admin test users.

**Comment by moio** [comment]
[User confirmed](https://github.com/rancher/rancher/pull/42865#issuecomment-1744798530) hotspot 1 has been fixed.

More research on successive hotspots will follow.

**Comment by moio** [comment]
User reported the problem was solved by restructuring RBAC in order to minimize the number of RoleBindings created.

Specifically, they could reduce from 40 to 13 restricted admins.

I am closing this issue now, and encourage anybody able to reproducibly observe substantially higher CPU/memory consumption when clusters are disconnected to open a new issue and CC me so that we can take a deeper look.

**Comment by git-ival** [comment]
Retested and confirmed the fix stops rancher from flooding with fleetworkspace rolebinding rewrites.

Repro:
1. Setup a reproduction environment on AWS using `scalability-tests` repo
	1. Rancher 2.7.6
	2. Use this branch (https://github.com/git-ival/scalability-tests/tree/disconnected_clusters_aws) for now to enable AWS setup
2. Setup Roles and restricted-admin Users by running `run_tests.mjs`
3. (Optional) Force downstream cluster(s) into an unavailable state
	1. Disable node replacement and delete underlying nodes, force cattle agent to be killed or use another method
4. Run `kubectl get -A --watch --watch-only rolebinding` and wait for at least 10 minutes
5. Observe that many rolebinding rewrites have occurred

Encountered periodic flooding of fleet rolebinding rewrites. The amount and frequency of rewrites was worse after forcing downstream clusters to be unavailable. See below for console output showing rolebinding rewrites
```
$ kubectl get -A --watch --watch-only rolebinding
NAMESPACE       NAME                                                   ROLE
   AGE
fleet-default   grb-9t8m5-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   31m
fleet-default   grb-dmh8n-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   20m
fleet-default   grb-q4xgl-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   34m
fleet-default   grb-7xl2d-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   17m
fleet-default   grb-9jbsk-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   30m
fleet-default   grb-s74cw-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   19m
fleet-default   grb-rdxl2-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   15m
fleet-default   grb-tfgqd-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   9m1s
fleet-default   grb-c2k68-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   35m
fleet-default   grb-mq9hl-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   13m
fleet-default   grb-njxb4-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   31m
fleet-default   grb-4lxcn-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   29m
fleet-default   grb-psrgb-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   36m
fleet-default   grb-x5wwh-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   20m
fleet-default   grb-qh9dd-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   36m
fleet-default   grb-2gg9d-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   38m
fleet-default   grb-sdslt-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   28m
fleet-default   grb-tcjm6-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   36m
fleet-default   grb-dsnvw-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   12m
fleet-default   grb-dmg6l-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   26m
fleet-default   grb-t79g7-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   17m
fleet-default   grb-zth9j-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   9m56s
fleet-default   grb-psgkj-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   38m
fleet-default   grb-qktx7-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   33m
fleet-default   grb-f2vr5-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   28m
fleet-default   grb-vll4j-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   28m
fleet-default   grb-ww47c-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   14m
fleet-default   grb-rlwc5-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   23m
fleet-default   grb-chrl6-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   28m
fleet-default   grb-5kqqr-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   13m
fleet-default   grb-zpvfk-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   38m
fleet-default   grb-rw467-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   26m
fleet-default   grb-pdwjj-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   35m
fleet-default   grb-g6lww-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   26m
fleet-default   grb-tr8qb-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   36m
fleet-default   grb-g65v5-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   9m22s
fleet-default   grb-74vld-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   38m
fleet-default   grb-sqqrx-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   24m
fleet-default   grb-zlm5g-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   28m
fleet-default   grb-dv9q2-fleetworkspace-restricted-admin-rb-cluster   ClusterRole/fleetworkspace-admin   15m
```

Validation:
1. Setup a validation environment on AWS using `scalability-tests` repo
	1. Rancher v2.7.6-debug-43051-1
	2. Use this branch (https://github.com/git-ival/scalability-tests/tree/disconnected_clusters_aws) for now to enable AWS setup
2. Setup Roles and restricted-admin Users by running `run_tests.mjs`
3. (Optional) Force downstream cluster(s) into an unavailable state
	1. Disable node replacement and delete underlying nodes, force cattle agent to be killed or use another method
4. Run `kubectl get -A --watch --watch-only rolebinding` and wait for at least 10 minutes
5. Observe that no rolebinding rewrites have occurred
6. Wait for another hour
7. Observe non-fleetworkspace rolebinding rewrites (as expected)

After waiting over an hour the following log lines were generated after step 5
```
$ kubectl get -A --watch --watch-only rolebinding
NAMESPACE             NAME                                              ROLE                                            AGE
cattle-fleet-system   helm-operation-7t48s-helm-installer-rolebinding   Role/helm-operation-7t48s-helm-installer-role   59m
cattle-fleet-system   helm-operation-7t48s-helm-installer-rolebinding   Role/helm-operation-7t48s-helm-installer-role   59m
cattle-fleet-system   helm-operation-zfg8q-helm-installer-rolebinding   Role/helm-operation-zfg8q-helm-installer-role   59m
cattle-fleet-system   helm-operation-zfg8q-helm-installer-rolebinding   Role/helm-operation-zfg8q-helm-installer-role   59m
cattle-fleet-system   helm-operation-4ph9z-helm-installer-rolebinding   Role/helm-operation-4ph9z-helm-installer-role   59m
cattle-fleet-system   helm-operation-4ph9z-helm-installer-rolebinding   Role/helm-operation-4ph9z-helm-installer-role   59m
cattle-fleet-system   helm-operation-8zdn8-helm-installer-rolebinding   Role/helm-operation-8zdn8-helm-installer-role   59m
cattle-fleet-system   helm-operation-8zdn8-helm-installer-rolebinding   Role/helm-operation-8zdn8-helm-installer-role   59m
cattle-fleet-system   helm-operation-4h55g-helm-installer-rolebinding   Role/helm-operation-4h55g-helm-installer-role   59m
cattle-fleet-system   helm-operation-4h55g-helm-installer-rolebinding   Role/helm-operation-4h55g-helm-installer-role   59m
cattle-system         helm-operation-mqbzf-helm-installer-rolebinding   Role/helm-operation-mqbzf-helm-installer-role   59m
cattle-system         helm-operation-mqbzf-helm-installer-rolebinding   Role/helm-operation-mqbzf-helm-installer-role   59m
```
